jsonstar/JsonStar.Schema.Ast.ToJsonSchema.fst

```diff
@@ -0,0 +1,197 @@
+    | Number opt -> { baseSchema with _type = Number (refineNumber ref opt); }
+    | Enum vs -> { baseSchema with _type = Enum (refineEnum ref vs); }
+    | Boolean -> baseSchema
+    // TODO: Apply other refinements
+    | _ -> baseSchema
+    
+and typedefToJsonSchema (x : AST.typedef) : Tot schema = 
+    let baseTyp = AST.Mktypedef?._base x in
+    let attrs = AST.Mktypedef?._attributes x in
+    let baseSchema = toJsonSchema baseTyp in
+    {
+        baseSchema with
+            description = AST.Mkvalue_attributes?._description attrs;
+            _default    = AST.Mkvalue_attributes?._defaultValue attrs;
+            readonly    = AST.Mkvalue_attributes?._readOnly attrs;
+            format      = Option.mapTot AST.format_type_to_string (AST.Mkvalue_attributes?._format attrs);
+    }
+
+and recordToJsonSchema (r : AST.record) : Tot schema = 
+    let props = 
+        JsonStar.Utils.list_map 
+            (AST.Mkrecord?._fields r)
+            (fun rf ->
+                match AST.Mkrecord_field?._typ rf with
+                // NOTE: OneOf record field is a special case where we generate an 
+                //       enum property and elsewhere a subschema is being put into deps
+                | AST.OneOf o -> begin
+                    match o with
+                    | AST.ByEnum n e _ -> n, (enumToJsonSchema e)
+                    end
+                | field_typ -> begin
+                    // TODO: Support attributes
+                    (AST.Mkrecord_field?._name rf), (toJsonSchema field_typ)
+                    end
+            ) 
+    in
+    let deps = 
+        JsonStar.Utils.list_choose
+            (AST.Mkrecord?._fields r)
+            (fun rf ->
+                match AST.Mkrecord_field?._typ rf with
+                | AST.OneOf o -> begin
+                    match o with
+                    | AST.ByEnum n e st -> begin
+                        let ds = 
+                            JsonStar.Utils.list_choose
+                                st
+                                (fun (v, so) -> 
+                                    match so with
+                                    | Some s -> Some (v, toJsonSchema s)
+                                    | None -> None
+                                )
+                        in Some (n, ds)
+                        end
+                    end
+                | _ -> None)
+    in
+    let req =  
+        JsonStar.Utils.list_choose 
+            (AST.Mkrecord?._fields r)
+            (fun rf ->
+                match AST.Mkrecord_field?._typ rf with
+                // NOTE: OneOf record field is a special case where we generate an 
+                //       enum property and it's always marked as required
+                | AST.OneOf (AST.ByEnum n _ _) -> Some n
+                | _ -> 
+                    if (AST.Mkrecord_field?._isOptional rf)
+                        then None
+                        else Some (AST.Mkrecord_field?._name rf)
+            )
```

Comment:
I don't like the fact that you are performing `JsonStar.Utils.list_choose` twice over the same argument `(AST.Mkrecord?._fields r)`. It looks like a code smell, but I don't understand what the code is doing so I can't judge. I'd try to change this so that there's only a single call to `list_choose`.